sw/runtime/src/start.c

```diff
@@ -50,9 +52,43 @@ static inline void snrt_init_tls() {
 
     snrt_cluster_hw_barrier();
 }
+#else
+// If dma is not enabled
+static inline void snrt_init_tls() {
+    extern volatile uint32_t __tdata_start, __tdata_end;
+    extern volatile uint32_t __tbss_start, __tbss_end;
+
+    size_t size;
+    volatile uint32_t tls_ptr;
+    size = (size_t)(&__tdata_end) - (size_t)(&__tdata_start);
+
+    // Every core initializes its own TLS
+    asm volatile("mv %0, tp" : "=r"(tls_ptr) : :);
+    memcpy((void*)tls_ptr, (const void*)&__tdata_start, size);
+
+    // Then initialize all other cores' .tdata sections from the DM
+    // core's. The offset between the TLS section of successive cores
+    // is defined in start.S
+    // Copy .tdata into all other cores' TLS regions
+    size_t tls_offset = (1 << SNRT_LOG2_STACK_SIZE) + 8;
+    for (int i = 1; i < snrt_cluster_core_num(); i++) {
+        memcpy((void*)(tls_ptr + i * tls_offset), (const void*)tls_ptr, size);
+    }
+
+    // Initialize all cores' .tbss sections
+    tls_ptr += size;
+    size = (size_t)(&__tbss_end) - (size_t)(&__tbss_start);
+    for (int i = 0; i < snrt_cluster_core_num(); i++) {
+        memset((void*)(tls_ptr + i * tls_offset), 0, size);
+    }
+
+    snrt_cluster_hw_barrier();
+}
+#endif
 #endif
```

Comment:
Instead of selecting between `memcpy` and `snrt_dma_start_1d` here, we should do it inside `snrt_dma_start_1d`. So that we can always use `snrt_dma_start_1d`, but if the DMA is not supported it will just be implemented as a regular `memcpy`. Same thing for `memset`.